src/appStorage.ts

```diff
@@ -0,0 +1,129 @@
+        default:
+          checksum = walletChecksum(accountPubKeyHex)
+      }
+    } else {
+      checksum = {ImagePart: '', TextPart: ''}
+    }
+    // missing checksum
+    walletMetaUpdate.checksum = checksum
+    await migrateAttribute('checksum', walletMetaUpdate)
+  }
+
+  // missing implementation id
+  if (!currentWalletMeta?.walletImplementationId && walletImplementationId) {
+    walletMetaUpdate.walletImplementationId = walletImplementationId
+    await migrateAttribute('walletImplementationId', walletMetaUpdate)
+  }
+
+  // resolving to a different network
+  if (currentWalletMeta?.networkId !== networkId) {
+    walletMetaUpdate.networkId = networkId
+    await migrateAttribute('networkId', walletMetaUpdate)
+  }
+
+  // missing isHW
+  if (!('isHW' in currentWalletMeta)) {
+    walletMetaUpdate.isHW = walletData != null && walletData?.isHW != null ? walletData.isHW : false
+    await migrateAttribute('isHW', walletMetaUpdate)
+  }
+
+  // missing isEasyConfirmationEnabled
+  if (!('isEasyConfirmationEnabled' in currentWalletMeta)) {
+    walletMetaUpdate.isEasyConfirmationEnabled = false
+    await migrateAttribute('isEasyConfirmationEnabled', walletMetaUpdate)
+  }
+
+  if (isWalletMeta(walletMetaUpdate)) return walletMetaUpdate as WalletMeta
+
+  Logger.error(`Invalid wallet meta `, JSON.stringify(walletMetaUpdate))
+  throw new Error(`Wallet meta stored is corrupted. ${JSON.stringify(walletMetaUpdate)}`)
+}
+
+export async function migrateWalletMetas(currentWalletMetas: Array<Partial<WalletMeta>>): Promise<Array<WalletMeta>> {
+  return Promise.all(currentWalletMetas.map(toShelleyWalletMeta))
+}
+
+function migrateAttribute(attr: keyof WalletMeta, walletMetaUpdated: Partial<WalletMeta>): Promise<void> {
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const id = walletMetaUpdated!.id
+  Logger.warn(`migrateAttribute::${attr} of wallet ${id}`)
+  return storage.write(`/wallet/${id}`, walletMetaUpdated)
+}
+
+function isWalletMeta(walletMeta?: WalletMeta | object | undefined): walletMeta is WalletMeta {
+  return (
+    // prettier-ignore
+    !!walletMeta &&
+    'id' in walletMeta 
+      && typeof walletMeta.id === 'string' &&
+    'name' in walletMeta 
+      && typeof walletMeta.name === 'string' &&
+    'networkId' in walletMeta 
+      && typeof walletMeta.networkId === 'number' &&
+    'isHW' in walletMeta 
+      && typeof walletMeta.isHW === 'boolean' &&
+    'isEasyConfirmationEnabled' in walletMeta 
+      && typeof walletMeta.isEasyConfirmationEnabled === 'boolean' &&
+    'checksum' in walletMeta 
+      && typeof walletMeta.checksum === 'object' &&
+    ('provider' in walletMeta 
+      && typeof walletMeta.provider === 'string'
+      || !('provider' in walletMeta)) &&
+    'walletImplementationId' in walletMeta 
+      && typeof walletMeta.walletImplementationId === 'string' 
+      && Object.values(WALLET_IMPLEMENTATION_REGISTRY).indexOf(walletMeta.walletImplementationId) > -1 &&
```

Comment:
.includes

Reply:
👍  copy and paste from the old validator